Implement "Go to Defintion" in the language server

darkdrag00n opened the pull request:
First part of #714 implementing the support for `Go to Definition` in Solang LSP.

I'll add support for other features such as `Go to type definition, ...` in follow-up PRs.

##### Small Demo
https://user-images.githubusercontent.com/119507009/209476628-24ec3af5-396b-4d64-8540-ff0395ec3a65.mp4

Signed-off-by: darkdrag00n <darkdrag00n@proton.me>

seanyoung reviewed (changes requested):
Looks very good! Some comments:
 - The `lookup_tbl` and `definition_tbl` should go into one struct, and passed around as one
 - Missing definitions for function definitions:  `Expression::InternalFunction`,`Expression::ExternalFunction`, `Expression::InternalFunctionCall`, Expression::ExternalFunctionCall`.
 - Missing contract `Expression::Constructor`
 - Event definition for `Statement::Emit`
 - Storage Definition for `Statement::Delete`
 - Fields of struct definition or event definition can be structs/enums too

darkdrag00n commented:
> Looks very good! Some comments:
> 
> * The `lookup_tbl` and `definition_tbl` should go into one struct, and passed around as one
> * Missing definitions for function definitions:  `Expression::InternalFunction`,`Expression::ExternalFunction`, `Expression::InternalFunctionCall`, Expression::ExternalFunctionCall`.
> * Missing contract `Expression::Constructor`
> * Event definition for `Statement::Emit`
> * Storage Definition for `Statement::Delete`
> * Fields of struct definition or event definition can be structs/enums too

1. Done, moved them in a struct named `Intelligence`. Please suggest alternative name if necessary
2. Works for me. Can you share the code on which it didn't work for you?
3. Done, added
4. Done, added
5. Done, added
6. Works for me with the nested structs/enums also. Same as point 2, please share the code on which it didn't work for you.

This is ready for another round of review from my end. Thanks!

seanyoung reviewed (changes requested):
Some test cases which aren't working:

```solidity


contract c {
    enum foo { bar, bar2 }
    struct S { foo f1; }

    function f() public {}

    function g() public {
        // Expression::InternalFunction
        function() internal x = f;
        // Expression::ExternalFunction
        function() external y = this.f;
        // Expression::New
        S[] z = new S[](2);
    }
}
```
